Authn-JWT Refactor

jvanderhoof opened the pull request:
This work is an effort to port the Authn-JWT functionality to the new Authenticator architecture.

### Desired Outcome

Refactor Authn-JWT to use the new authenticator architecture defined in the new OIDC authenticator.

### Implemented Changes

The authenticator has been completely ported, with all previous integration tests passing.  I've added a [Readme](https://github.com/cyberark/conjur/blob/b563eb9bb1e116dcbeebd26ce72d02689102b066/app/domain/authentication/Readme.md) which overviews the architecture and interfaces. 

### Connected Issue/Story

N/A

### Definition of Done

#### Changelog

- [ ] The CHANGELOG has been updated, or
- [x] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [x] This PR includes new unit and integration tests to go with the code
  changes, or
- [ ] The changes in this PR do not require tests

#### Documentation

- [x] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]
- [ ] This PR does not require updating any documentation

#### Behavior

- [ ] This PR changes product behavior and has been reviewed by a PO, or
- [ ] These changes are part of a larger initiative that will be reviewed later, or
- [x] No behavior was changed with this PR

#### Security

- [x] Security architect has reviewed the changes in this PR,
- [ ] These changes are part of a larger initiative with a separate security review, or
- [ ] There are no security aspects to these changes

rafis3 reviewed (commented):
Thanks for this design @jvanderhoof.
It's really visible in your implementation, of how much the different steps of the authenticator logic is organized and easier to understand. I added a few comments and questions, would love for you to take a look.

jvanderhoof commented on app/domain/authentication/Readme.md:
```diff
@@ -0,0 +1,227 @@
+# @param [String] request_body - authentication request body
+# @param [Hash] parameters - authentication request parameters
+#
+# @return something suitable for identifying a Conjur Role (usually a String or Hash)
```
In the case of the JWT Authenticator, we return the relevant part of the JWT token. Initially I was thinking is is because we perform some extraction in the Identity Resolution phase. In the OIDC Authenticator, we extract the desired lookup attribute from the claim (and return that). 

Until you asked the question, I didn't think about the fact that the ["pre-filter" logic](https://github.com/cyberark/conjur/blob/049bc4c13b3bc30fef8355312caa268b765036dc/app/domain/authentication/authn_jwt/v2/resolve_identity.rb#L14-L45) probably should live in the Strategy (and Strategies need to return a single identity to be mapped to a Conjur Role). This allows us to keep the implementation details (JWT/OIDC/AWS IAM, etc.) in the Strategy.

This change also enables the above refactor (streaming role match) to be significantly simpler.

jvanderhoof replied:
Scratch part of this above statement.  There is key information in the JWT that needs to be available in the Identity Resolution phase (ex annotation matching).

jvanderhoof committed:
Authn-JWT refactor

This commit migrates the existing authn-jwt functionality to the new Strategy/ResolveIdentity architecture.

jvanderhoof committed:
Interface updates to Authn-OIDC code redirect authentication flow

This commit updates the previously implemented authn-oidc workflow to adhere to the small changes in interface
defined in the authn-jwt refactor

jvanderhoof committed:
Removes previous authn-jwt code

This commit removes the previous authn-jwt implementation and unit tests.

jvanderhoof committed:
Fixes Authenticator list endpoint

jvanderhoof committed:
Refactored for code climate

jvanderhoof committed:
Refactor resolver based on CC

jvanderhoof committed:
Additional refactoring

jvanderhoof committed:
Additional refactoring

jvanderhoof committed:
Refactor on status handler

jvanderhoof committed:
Refactor method length

jvanderhoof committed:
Refactor

jvanderhoof committed:
More refactors

jvanderhoof committed:
Refactoring

jvanderhoof committed:
Refactor

jvanderhoof committed:
Refactor

jvanderhoof committed:
Refactoring

jvanderhoof committed:
Pulls conjur-rack gem directly into Conjur

Conjur is currently the only project which uses the conjur-rack gem. This PR brings it into Conjur to simplify
the authorization process.

jvanderhoof committed:
Adds support for authenticating with the AWS STS regional endpoint

This commit:
  - Adds support for regional STS endpoints.
  - Improves log messaging for failed requests.
  - Adds unit tests for all authenticator functionality.

jvanderhoof committed:
Removes a Git call from gemspec

This commit:
- Replaces a call to `git` with the equivalent Ruby call. This prevents a Git permission error.
- Addresses all Rubocop warnings.

jvanderhoof committed:
CC refactors

jvanderhoof committed:
CC fixes

jvanderhoof committed:
additional cc cleanup

jvanderhoof committed:
More fixes

jvanderhoof committed:
More cc fixes

jvanderhoof committed:
Merge branch 'master' into authn-jwt-refactor

szh commented:
@jvanderhoof Can this PR be closed?
